File: implement-shell-tools/wc/wc.mjs
import { promises as fs } from "node:fs";
import { program } from "commander";


const NUMBER_PADDING = 8;
const ERROR_PREFIX = "wc.mjs:"
const TOTAL_PATH = "total";


program
  .name("word-count")
  .description("word, line, and character count")
  .option("-l", "The number of lines in each input file is written to the standard output.")
  .option("-w", "The number of words in each input file is written to the standard output.")
  .option("-c", "The number of characters in each input file is written to the standard output.")
  .argument("[path...]", "path to file.");

program.parse();

const argv = program.args;
const options = program.opts();


start();

async function start() {
  const data = await collectData();
  const output = formatOutput(data);

  printOutput(output);
}

async function collectData() {
  const data = [];

  if (!argv.length) {
    const content = await getStdinContent();
    data.push(collectFileData({}, content));
  } else {
    for (const path of argv) {
      const datum = {};
      try {
        const stat = await fs.stat(path)
        if (stat.isDirectory()) {
          datum["s"] = "d";
        }
        if (stat.isFile()) {
          const content = (await fs.readFile(path)).toString();
          collectFileData(datum, content);
        }
      } catch (error) {
        datum["s"] = "e";
      }
      datum["p"] = path;
      data.push(datum);
    }
  }
  return data;
}

async function getStdinContent() {
  return new Promise((resolve) => {
    let content = "";

    process.stdin.on("data", input => {
      content += input;
    });
    process.stdin.on("close", () => {
      resolve(content);
    });
  });
}

function collectFileData(datum, content) {
  const newLineCharacters = content.match(/\n/g);

  datum["s"] = "f";
  datum["l"] = newLineCharacters ? newLineCharacters.length : 0;
  datum["w"] = content.trim().split(/\s+/).length;
  datum["c"] = content.length;

  return datum;
}

function formatOutput(data) {
  const errors = [];
  const files = [];

  for(const datum of data) {
    switch(datum.s) {
      case "d":
        errors.push(`${ERROR_PREFIX} ${datum.p}: Is a directory`);
        break;
      case "f":
        files.push(formatFileOutput(datum));
        break;
      default:
        errors.push(`${ERROR_PREFIX} ${datum.p}: open: No such file or directory`);
        break;
    }
  }
  
  if (data.length > 1) {
    files.push(formatTotalOutput(data));
  }
  
  return errors.concat(files);
}

function formatFileOutput(datum) {
  let result = "";
  
  if (!options.l && !options.w && !options.c) {
    result = `${formatOutputNumber(datum.l)}${formatOutputNumber(datum.w)}${formatOutputNumber(datum.c)}`;
  } else {
    if (options.l) {
      result = `${result}${formatOutputNumber(datum.l)}`;
    }
    if (options.w) {
      result = `${result}${formatOutputNumber(datum.w)}`;
    }
    if (options.c) {
      result = `${result}${formatOutputNumber(datum.c)}`;
    }
  }

  if (datum.p) {
    result = `${result} ${datum.p}`;
  }

  return result;
}

function formatTotalOutput(data) {
  const total = {s: "t", l: 0, w: 0, c: 0, p: TOTAL_PATH};

  for (const datum of data) {
    if (datum.s === "f") {
      total.l += datum.l;
      total.w += datum.w;
      total.c += datum.c;
    }
  }
  
  return formatFileOutput(total);
}

function formatOutputNumber(number) {
  return number.toString().padStart(NUMBER_PADDING, " ");
}

function printOutput(output) {
  for (const string of output) {
    if (string.startsWith(ERROR_PREFIX)) {
      console.error(string);
    } else {
      console.log(string);
    }
  }
}

File: implement-shell-tools/wc/wc.py
import os
import sys
import argparse

NUMBER_PADDING = 8
ERROR_PREFIX = "wc.mjs:"
TOTAL_PATH = "total"


parser = argparse.ArgumentParser(
  prog = "word-count",
  description = "word, line, and character count"
)

parser.add_argument("path", help = "path to file.", nargs = "*")
parser.add_argument("-l", action = "store_true", help = "The number of lines in each input file is written to the standard output.")
parser.add_argument("-w", action = "store_true", help = "The number of words in each input file is written to the standard output.")
parser.add_argument("-c", action = "store_true", help = "The number of characters in each input file is written to the standard output.")

args = parser.parse_args()


def start():
  data = collectData()
  output = formatOutput(data)
  printOutput(output)

def collectData():
  data = []
  if not len(args.path):
    content = sys.stdin.read()
    data.append(collectFileData({}, content))
  else:
    for path in args.path:
      datum = {}
      if (os.path.isdir(path)):
        datum['s'] = "d"
      elif (os.path.isfile(path)):
        with open(path, "r") as file: 
          content = file.read()
        collectFileData(datum, content)
      else:
        datum['s'] = "e"
      datum['p'] = path
      data.append(datum)

  return data

def collectFileData(datum, content):
  datum['s'] = "f"
  datum['l'] = content.count("\n")
  datum['w'] = len(content.strip().split())
  datum['c'] = len(content)

  return datum

def formatOutput(data):
  errors = []
  files = []

  for datum in data:
    if (datum.get('s') == "d"):
      errors.append(f"{ERROR_PREFIX} {datum.get('p')}: Is a directory")
    elif (datum.get('s') == "f"):
      files.append(formatFileOutput(datum))
    else:
      errors.append(f"{ERROR_PREFIX} {datum.get('p')}: open: No such file or directory")

  if (len(data) > 1):
    files.append(formatTotalOutput(data))
  
  return errors + files

def formatFileOutput(datum):
  result = ""
  if (not args.l and not args.w and not args.c):
    result = f"{formatOutputNumber(datum.get('l'))}{formatOutputNumber(datum.get('w'))}{formatOutputNumber(datum.get('c'))}"
  else:
    if (args.l):
      result = f"{result}{formatOutputNumber(datum.get('l'))}"
    if (args.w):
      result = f"{result}{formatOutputNumber(datum.get('w'))}"
    if (args.c):
      result = f"{result}{formatOutputNumber(datum.get('c'))}"

  if (datum.get('p')):
    result = f"{result} {datum['p']}"
  
  return result

def formatTotalOutput(data):
  total = {"s": "t", "l": 0, "w": 0, "c": 0, "p": TOTAL_PATH};
  
  for datum in data:
    if (datum.get('s') == "f"):
      total['l'] += datum['l']
      total['w'] += datum['w']
      total['c'] += datum['c']
  
  return formatFileOutput(total)

def formatOutputNumber(number):
  return str(number).rjust(NUMBER_PADDING)

def printOutput(output):
  for string in output:
    if (string.startswith(ERROR_PREFIX)):
      print(string, file = sys.stderr)
    else:
      print(string)

start()
